### ISSUE TITLE
Goto +foo broken

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6811]
Migrated from [rt.perl.org#24108](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=24108) (status was 'resolved')

Searchable as RT24108$

**Comment by p5pRT** [comment]
### From @Juerd

#### Created by @Juerd



goto \+foo doesn't properly go to the thing returned by foo\(\)\.

foo\(\) is called\, but goto uses a reference to the value that it returns\.

2;0 juerd@&#8203;ouranos&#8203;:\~$ perl \-le'sub foo \{print "x"\} goto \+foo'
x
Can't find label SCALAR\(0x812139c\) at \-e line 1\.
2;255 juerd@&#8203;ouranos&#8203;:\~$ perl \-MO=Deparse \-le'sub foo \{print "x"\} goto \+foo'
BEGIN \{ $/ = "\\n"; $\\ = "\\n"; \}
sub foo \{
&nbsp;    print 'x';
\}
goto \\foo ;
\-e syntax OK

Note the \\ before foo\.

XMATH knows more than I do&#8203;:

23&#8203;:36 \<@&#8203;xmath> Juerd&#8203;: notice the difference of 64 in op\_flags\.\. that's the 
&nbsp;               OPf\_STACKED\-flag\, which means the sub will get its arguments via stack 
&nbsp;               \(rather than borrowing caller's @&#8203;\_\, which is what &foo does\)
23&#8203;:37 \<@&#8203;xmath> now\, goto checks if its argument is an ENTERSUB\, and if assumes it's goto 
&nbsp;               &NAME
23&#8203;:38 \<@&#8203;xmath> it'll pass the argument to an optree\-mongler\, telling it to turn the arg 
&nbsp;               into a ref\-able thing
23&#8203;:40 \<@&#8203;xmath> Juerd&#8203;: it should either include OPf\_STACKED in the test that decides 
&nbsp;               between goto &NAME and goto EXPR
23&#8203;:40 \<@&#8203;xmath> Juerd&#8203;: or use a different optree mongler that doesn't care about 
&nbsp;               OPf\_STACKED
23&#8203;:40 \<@&#8203;xmath> the former is probably the better solution
23&#8203;:41 \<@&#8203;xmath> well\, I'm not sure if you can consider it a bug
23&#8203;:41 \<@&#8203;xmath> and the behavior can be useful
23&#8203;:41 \<@&#8203;Juerd> Of course you can
23&#8203;:41 \<@&#8203;Juerd> Useful how?
23&#8203;:42 \<@&#8203;xmath> I'm thinking very hard about that right now\.\. my thoughts were reaching 
&nbsp;               towards making foo an lvalue sub and return a var that is actually blessed 
&nbsp;               so that \\var will be an object\, then use overloading to make it stringify 
&nbsp;               into a label
23&#8203;:42 \<@&#8203;xmath> but I'm not sure if that would work ;\)
23&#8203;:42 \<@&#8203;Juerd> Dude\, just return that string label then
23&#8203;:43 \<@&#8203;xmath> hehe
23&#8203;:46 \<@&#8203;xmath> Juerd&#8203;: actually\.\.\.\. in Perl\_newLOOPEX\, change
23&#8203;:46 \<@&#8203;xmath>         if \(label\->op\_type == OP\_ENTERSUB\)
23&#8203;:46 \<@&#8203;xmath> to&#8203;:
23&#8203;:46 \<@&#8203;xmath>         if \(label\->op\_type == OP\_ENTERSUB && \!\(label\->op\_flags & 
&nbsp;               OPf\_STACKED\)\)
23&#8203;:46 \<@&#8203;xmath> that's the fix :\-\)
23&#8203;:47 \<@&#8203;xmath> Juerd&#8203;: just mention that fix in the report

This fix has not been tested\.

Note\, if the sub is not "foo" but "x"&#8203;:

2;255 juerd@&#8203;ouranos&#8203;:\~$ perl \-le'sub foo \{print "x"\} goto foo'
Can't find label foo at \-e line 1\.
2;255 juerd@&#8203;ouranos&#8203;:\~$ perl \-le'sub x \{print "x"\} goto x'
x
Can't find label SCALAR\(0x812139c\) at \-e line 1\.

\(Difference with previous snippet is "goto \+foo" => "goto foo"\)

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.8.0:

Configured by root at Sat Aug  9 18:22:36 CEST 2003.

Summary of my perl5 (revision 5.0 version 8 subversion 0) configuration:
  Platform:
    osname=linux, osvers=2.4.21-rc2, archname=i686-linux
    uname='linux ouranos 2.4.21-rc2 #3 sat may 17 00:09:50 cest 2003 i686 amd athlon(tm) processor authenticamd gnulinux '
    config_args='-des -Darchname=i686-linux -Dcc=gcc -Dprefix=/usr -Dvendorprefix=/usr -Dsiteprefix=/usr -Dlocincpth=  -Doptimize=-O2 -march=athlon-tbird -pipe -fomit-frame-pointer  -Duselargefiles -Dd_dosuid -Dd_semctl_semun -Dscriptdir=/usr/bin -Dman3ext=3pm -Dcf_by=Gentoo -Ud_csh -Di_gdbm -Di_db -Di_ndbm'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=define d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='gcc', ccflags ='-DPERL5 -fno-strict-aliasing -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -march=athlon-tbird -pipe -fomit-frame-pointer ',
    cppflags='-DPERL5 -DPERL5 -fno-strict-aliasing'
    ccversion='', gccversion='3.2.3 20030422 (Gentoo Linux 1.4 3.2.3-r1, propolice)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lpthread -lnsl -lndbm -lgdbm -ldb -ldl -lm -lc -lcrypt -lutil
    perllibs=-lpthread -lnsl -ldl -lm -lc -lcrypt -lutil
    libc=/lib/libc-2.3.2.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.3.2'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.8.0:
    /etc/perl
    /usr/lib/perl5/site_perl/5.8.0/i686-linux
    /usr/lib/perl5/site_perl/5.8.0
    /usr/lib/perl5/site_perl/5.6.1
    /usr/lib/perl5/site_perl
    /usr/lib/perl5/vendor_perl/5.8.0/i686-linux
    /usr/lib/perl5/vendor_perl/5.8.0
    /usr/lib/perl5/vendor_perl
    /usr/lib/perl5/5.8.0/i686-linux
    /usr/lib/perl5/5.8.0
    /usr/local/lib/site_perl
    /usr/lib/perl5/site_perl/5.6.1
    .


Environment for perl v5.8.0:
    HOME=/home/juerd
    LANG (unset)
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/bin:/usr/bin:/bin:/usr/bin:/bin:/usr/bin:/usr/local/bin:/opt/bin:/usr/i686-pc-linux-gnu/gcc-bin/3.2:/usr/X11R6/bin:/opt/eclipse:/opt/blackdown-jdk-1.4.1/bin:/opt/blackdown-jdk-1.4.1/jre/bin:/usr/qt/3/bin:/usr/qt/3-embedded/bin:/usr/kde/3.1/bin:/usr/kde/3.1/bin:/usr/games/bin:/opt/vmware/bin:/var/qmail/bin
    PERL_BADLANG (unset)
    SHELL=/bin/bash

Juerd

```
</details>

**Comment by p5pRT** [comment]
### From @rgs

Juerd \(via RT\) wrote&#8203;:
> 
> goto \+foo doesn't properly go to the thing returned by foo\(\)\.
> 
> foo\(\) is called\, but goto uses a reference to the value that it returns\.

This should be fixed in bleadperl by change \#21425\, thanks\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'new' to 'resolved'
